test/e2e/apps/job.go

```diff
@@ -111,18 +110,20 @@ var _ = SIGDescribe("Job", func() {
 
 	// Pods sometimes fail, but eventually succeed, after pod restarts
 	ginkgo.It("should run a job to completion when tasks sometimes fail and are not locally restarted", func() {
+		// One failure, then a success, no local restarts.
+		// We can't use the random failure approach, because JobController
+		// will throttle frequently failing Pods of a given Job, ramping
+		// up to 6 minutes between restarts, making test timeout due to
+		// successive failures.
+		// Instead, we force the Job's Pods to be scheduled to a single Node
```

Comment:
I just want to confirm that making this change doesn't violate the spirit of the test:

The test description says `"it should run a job to completion when tasks sometimes fail and are not locally restarted."` Am I understanding correctly that the `not locally restarted` component refers to the `v1.RestartPolicyNever`? Or does `not locally restarted` mean that we want to ensure the pod is actually running on different nodes?

If the former is true, I agree with the approach you've taken. However, if the later is true, I worry that the changes made here may be violating the spirit of the test.

Reply:
According to the difference of the two tests that only `RestartPolicy` is changed, I think it's the former, I guess "locally restart" here means kubelet recreates the failed containers of the Pod locally:
`should run a job to completion when tasks sometimes fail and are locally restarted`
`should run a job to completion when tasks sometimes fail and are not locally restarted`

If my understanding is wrong, I guess we will have to use the http approach to conduct the Pod succeed and fail like my first approach did, but it looks not clear like the current one @liggitt suggested.

Reply:
the containers are not locally restarted, they fail and drive completely new pods/containers